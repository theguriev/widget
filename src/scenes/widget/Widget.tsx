import React, { FC, useState, useCallback } from 'react';
import { Button } from '@/components';
import XIcon from '@/assets/icons/x.svg';
import SmsIcon from '@/assets/icons/sms.svg';
import InstagramIcon from '@/assets/icons/instagram.svg';
import MessengerIcon from '@/assets/icons/facebook-messenger.svg';
import { applyHook, joinClasses } from '@/utils';
import { useEvent } from '@/hooks';
import styles from './styles.module.css';
import { WidgetProps, StartButtonProps } from './types';

export const useWidget = () => {
  const [isOpen, setIsOpen] = useState(false);
  const [isMinimized, seIsMinimized] = useState(false);

  const handleToggle = () =>
    setIsOpen((current) => {
      window.postMessage('toggle', '*');
      window.parent.postMessage('toggle', '*');
      return !current;
    });

  const handleMessage = useCallback(
    (event: MessageEvent) => {
      const { data } = event;
      if (data === 'scroll') {
        seIsMinimized(true);
      }
    },
    [seIsMinimized]
  );

  useEvent('message', handleMessage);

  return {
    isOpen,
    isMinimized,
    handleToggle,
  };
};

const StartButton: FC<StartButtonProps> = ({ isMinimized, handleToggle }) => {
  const content = isMinimized ? (
    <>
      <div className={styles.dot} />
      <div className={styles.dot} />
      <div className={styles.dot} />
    </>
  ) : (
    <>
      <SmsIcon />
      <MessengerIcon />
      <InstagramIcon />
      Message us
    </>
  );
  return (
    <Button
      variant={isMinimized ? 'circle' : 'list'}
      onClick={handleToggle}
      className={joinClasses.apply(null, [
        styles.startButton,
        isMinimized ? styles.dots : '',
      ])}
    >
      {content}
    </Button>
  );
};

const WidgetRepresentation: FC<WidgetProps> = ({
  isOpen,
  isMinimized,
  handleToggle,
}) => (
  <div className={styles.widget}>
    <div className={joinClasses(styles.list, isOpen ? styles.open : '')}>
      <Button variant="list">
        <SmsIcon />
        SMS
      </Button>
      <Button variant="list">
        <MessengerIcon />
        Messenger
      </Button>
      <Button variant="list">
        <InstagramIcon />
        Instagram
      </Button>
    </div>
    {isOpen ? (
      <Button onClick={handleToggle}>
        <XIcon />
      </Button>
    ) : (
      <StartButton handleToggle={handleToggle} isMinimized={isMinimized} />
    )}
  </div>
);

export const Widget = applyHook(WidgetRepresentation, useWidget);
