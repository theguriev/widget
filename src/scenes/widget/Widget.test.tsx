import * as React from 'react';
import { act, render, renderHook } from '@testing-library/react';

import { Widget, useWidget } from './Widget';

const handleToggle = () => {};

describe('Widget', () => {
  it('renders Widget not minimized and closed', () => {
    const { asFragment } = render(<Widget handleToggle={handleToggle} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Widget minimized and closed', () => {
    const { asFragment } = render(
      <Widget isMinimized handleToggle={handleToggle} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Widget not minimized and open', () => {
    const { asFragment } = render(
      <Widget isOpen handleToggle={handleToggle} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Widget minimized and open', () => {
    const { asFragment } = render(
      <Widget isOpen isMinimized handleToggle={handleToggle} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

const listeners = new Map<string, Set<(event) => void>>();

describe('useWidget', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'addEventListener', {
      value: (type, listener) => {
        if (type !== 'error') {
          listeners.set(type, listeners.get(type) || new Set());
          listeners.get(type).add(listener);
        }
      },
    });
    Object.defineProperty(window, 'removeEventListener', {
      value: (type, listener) => {
        if (type !== 'error') {
          listeners.set(type, listeners.get(type) || new Set());
          listeners.get(type).delete(listener);
        }
      },
    });
  });

  test('handleToggle', () => {
    const { result } = renderHook(() => useWidget());
    expect(result.current.isOpen).toBe(false);
    act(() => {
      result.current.handleToggle();
    });
    expect(result.current.isOpen).toBe(true);
    act(() => {
      result.current.handleToggle();
    });
    expect(result.current.isOpen).toBe(false);
  });

  test('handleMessage', () => {
    const { result } = renderHook(() => useWidget());
    expect(result.current.isMinimized).toBe(false);
    const callbacks = listeners.get('message');
    expect(callbacks.size).toBe(1);
    act(() => {
      callbacks.forEach((callback) => {
        callback({ data: 'scroll' });
      });
    });
    expect(result.current.isMinimized).toBe(true);
  });
});
