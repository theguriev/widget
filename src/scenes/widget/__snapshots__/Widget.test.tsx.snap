// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget renders Widget minimized and closed 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="list"
    >
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgSms"
        />
        SMS
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgFacebookMessenger"
        />
        Messenger
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgInstagram"
        />
        Instagram
      </button>
    </div>
    <button
      class="circle startButton dots"
      type="button"
    >
      <div
        class="dot"
      />
      <div
        class="dot"
      />
      <div
        class="dot"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Widget renders Widget minimized and open 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="list open"
    >
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgSms"
        />
        SMS
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgFacebookMessenger"
        />
        Messenger
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgInstagram"
        />
        Instagram
      </button>
    </div>
    <button
      class="circle"
      type="button"
    >
      <svg
        data-file-name="SvgX"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Widget renders Widget not minimized and closed 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="list"
    >
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgSms"
        />
        SMS
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgFacebookMessenger"
        />
        Messenger
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgInstagram"
        />
        Instagram
      </button>
    </div>
    <button
      class="list startButton"
      type="button"
    >
      <svg
        data-file-name="SvgSms"
      />
      <svg
        data-file-name="SvgFacebookMessenger"
      />
      <svg
        data-file-name="SvgInstagram"
      />
      Message us
    </button>
  </div>
</DocumentFragment>
`;

exports[`Widget renders Widget not minimized and open 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="list open"
    >
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgSms"
        />
        SMS
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgFacebookMessenger"
        />
        Messenger
      </button>
      <button
        class="list"
        type="button"
      >
        <svg
          data-file-name="SvgInstagram"
        />
        Instagram
      </button>
    </div>
    <button
      class="circle"
      type="button"
    >
      <svg
        data-file-name="SvgX"
      />
    </button>
  </div>
</DocumentFragment>
`;
